CVAT 3D Milestone-5

manasars opened the pull request:
CVAT-3D-Milestone5 : Implement cuboid operations in right side bar and save annotations
Changes include:
Implemented displaying list of annotated objects in right side bar.
Implemented Switch lock property, switch hidden, pinned, occluded property of objects.
Implemented remove and save annotations
Implemented Appearance tab to change opacity and outlined borders.

Test Cases:
Manual Unit testing done locally.
Existing test cases work as expected.
System Test cases will be shared.

[x ] I submit my changes into the develop branch

[ x] I submit my code changes under the same MIT License that covers the project.

bsekachev commented:
Hi @manasars 

I checked the code, but probably will do it one more time after fixing the comments. 
Also checked the patch and here some comments about it:

- [ ] How to deactivate a shape? In 2D canvas you can click anywhere out of shape and activated shape will be deactivated. In 3D canvas it does not work. Need to fix. 
- [ ] Redraw shape does not work (Shift + N when an object is activated). It just creates a new object. Need to fix.
- [ ] Escape does not work when drawing (It should prevent drawing and put canvas to IDLE state). Need to fix.
- [ ] When we draw using N hotkey, wrong icon activated on controls-panel. Need to fix.
- [ ] What is difference between move control and cursor control in 3D canvas? Do we need both? 
- [ ] Extra space in the bottom of appearance block. Probably its height should be corrected because its content less then for 2D canvas. 
- [ ] Fit does not work by dblclick. Probably it should affect our users, because they used to use dblclick to fit 2D canvas. It does not work for 3D canvas. @nmanovic , what do you think?
- [ ] Need to update README.md (specify actual API of canvas there)
- [ ] When object locked cannot I select it (it does not appear on top/side/front views after click). Need to fix. 
Steps to reproduce:
   - Open a job
   - Create a cuboid by N (not click the object after it is created)
   - Lock the cuboid in the side bar
   - Try to click the cuboid now
- [ ] Selected opacity does not work properly. Please, look at gif (on the gif I tried to click objects a lot):
![Peek 2021-04-13 22-00](https://user-images.githubusercontent.com/40690378/114606208-acf0da00-9ca3-11eb-95f8-693b82df11ac.gif)
- [ ] When I go to fullscreen, I get top/side/front views offset
![Screenshot from 2021-04-13 22-02-41](https://user-images.githubusercontent.com/40690378/114606596-21c41400-9ca4-11eb-9f2d-3c5d98d4a54a.png)
- [ ] What have you decided about preview for 3D task? Before you said it will be solved in the future PRs. 
- [ ] Points on top/side/front views are almost invisible (maybe at least try to do them white? Just black and blue are not really contrast colors). Try to create a task with cvat/cvat/apps/engine/tests/assets/test_pointcloud_pcd.zip


Please, check checkbox if the problem is solved or if you believe it is not an issue, explain why.

nmanovic commented:
@jayraj-solanki , tests are failed. Please look.

bsekachev commented:
@jayraj-solanki @manasars 

Please, click the checkboxes above if you believe the corresponding issues were solved

jayraj-solanki commented:
Hi @bsekachev
* [x]  How to deactivate a shape? In 2D canvas you can click anywhere out of shape and activated shape will be deactivated. In 3D canvas it does not work. Need to fix.
* [x]  Redraw shape does not work (Shift + N when an object is activated). It just creates a new object. Need to fix.
 * [x]  Escape does not work when drawing (It should prevent drawing and put canvas to IDLE state). Need to fix.
* [x]  When we draw using N hotkey, wrong icon activated on controls-panel. Need to fix.
* [x]  What is difference between move control and cursor control in 3D canvas? Do we need both?
* [x]  Extra space in the bottom of appearance block. Probably its height should be corrected because its content less then for 2D canvas.
* [x]  Need to update README.md (specify actual API of canvas there)
* [x]  When object locked cannot I select it (it does not appear on top/side/front views after click). Need to fix.
* [x]  Selected opacity does not work properly. 
* [x]  When I go to fullscreen, I get top/side/front views offset

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3d.ts:
```diff
@@ -16,14 +22,18 @@ const Canvas3dVersion = pjson.version;
 interface Canvas3d {
     html(): ViewsDOM;
-    setup(frameData: any): void;
+    setup(frameData: any, objectStates: any[], zLayer?: number): void;
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dModel.ts:
```diff
@@ -39,14 +44,30 @@ export enum MouseInteraction {
 }
 
+export interface FocusData {
+    clientID: any;
```
Converted to string

bsekachev replied:
This is probably number, not string.

manasars replied:
In canvas3D it will remain string since we are using data structure of Three.js

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -251,6 +514,7 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
         this.views.top.scene.add(points.clone());
         this.views.side.scene.add(points.clone());
         this.views.front.scene.add(points.clone());
+        model.updateObject();
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -188,17 +357,107 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                    }
+                }
+            });
+            this.model.updateObject();
```
Removed

jayraj-solanki commented on cvat-ui/src/reducers/settings-reducer.ts:
```diff
@@ -299,6 +299,11 @@ export default (state = defaultState, action: AnyAction): SettingsState => {
+                shapes: {
+                    ...defaultState.shapes,
+                    opacity: (job.task.dimension === '3d') ? 40 : 3,
+                    selectedOpacity: (job.task.dimension === '3d') ? 60 : 30,
```
Done

jayraj-solanki commented on cvat-ui/src/components/annotation-page/top-bar/right-group.tsx:
```diff
@@ -152,6 +153,9 @@ function RightGroup(props: Props): JSX.Element {
                         } else {
                             window.document.documentElement.requestFullscreen();
                         }
+                        if (canvasInstance instanceof Canvas3d) {
```
Done

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/standard3D-workspace.tsx:
```diff
@@ -8,12 +8,21 @@ import Layout from 'antd/lib/layout';
+import ObjectsListContainer from '../../../containers/annotation-page/standard-workspace/objects-side-bar/objects-list';
+import PropagateConfirmContainer from '../../../containers/annotation-page/standard-workspace/propagate-confirm';
+import CanvasContextMenuContainer from '../../../containers/annotation-page/canvas/canvas-context-menu';
+import CanvasPointContextMenuComponent from '../canvas/canvas-point-context-menu';
```
Done

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/controls-side-bar/draw-cuboid-control.tsx:
```diff
@@ -40,15 +43,17 @@ function DrawPolygonControl(props: Props): JSX.Element {
+    return disabled ? (
+        <Icon className='cvat-draw-cuboid-control cvat-disabled-canvas-control' component={CubeIcon} />
+    ) : (
+        <CustomPopover
```
File deleted

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/controls-side-bar/cursor-control.tsx:
```diff
@@ -7,11 +7,12 @@ import Icon from '@ant-design/icons';
 import { ActiveControl } from 'reducers/interfaces';
-import { Canvas3d as Canvas } from 'cvat-canvas3d-wrapper';
+import { Canvas3d } from 'cvat-canvas3d-wrapper';
+import { Canvas } from 'cvat-canvas-wrapper';
```
File deleted

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/controls-side-bar/move-control.tsx:
```diff
@@ -7,25 +7,35 @@ import Icon from '@ant-design/icons';
+                        canvasInstance.cancel();
+                        canvasInstance.dragCanvas(true);
+                    }
+                }}
```
File deleted

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/controls-side-bar/controls-side-bar.tsx:
```diff
@@ -5,37 +5,112 @@
+import ControlVisibilityObserver from '../../standard-workspace/controls-side-bar/control-visibility-observer';
+import { Props as MoveControlProps } from '../../standard-workspace/controls-side-bar/move-control';
+import { Props as CursorControlProps } from '../../standard-workspace/controls-side-bar/cursor-control';
+import { Props as DrawCuboidControlProps } from '../../standard-workspace/controls-side-bar/draw-cuboid-control';
```
Done

jayraj-solanki commented on cvat-canvas3d/src/typescript/cuboid.ts:
```diff
@@ -2,19 +2,139 @@
+        // @ts-ignore
+        this.side.material.opacity = opacity / 100;
+        // @ts-ignore
+        this.front.material.opacity = opacity / 100;
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/cuboid.ts:
```diff
@@ -2,19 +2,139 @@
+    rotationHelper.name = 'rotationHelper';
+    switch (viewType) {
+        case 'top':
+            // @ts-ignore
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/cuboid.ts:
```diff
@@ -2,19 +2,139 @@
+
+            rotationHelper.position.set(
+                instance.position.x,
+                // @ts-ignore
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -305,27 +569,24 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                    const object = this.views.perspective.scene.getObjectByName(
+                        this.model.data.focusData.clientID,
+                    );
+                    // @ts-ignore
```
Removed

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -188,17 +357,107 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                    cuboid.setPosition(object.points[0],
+                        object.points[1], object.points[2]);
+                }
+            }
```
Function setupObjects implemented

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -188,17 +357,107 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                        color = object.color;
+                    } else {
+                        // eslint-disable-next-line prefer-destructuring
+                        color = object.group.color;
```
Done

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -188,17 +357,107 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+            const { clientID } = this.model.data.activeElement;
+            Object.keys(this.views).forEach((view: string): void => {
+                const viewType = this.views[view as keyof Views];
+                // @ts-ignore
```
Done

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -188,17 +357,107 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                const viewType = this.views[view as keyof Views];
+                // @ts-ignore
+                const object = viewType.scene.getObjectByName(clientID);
+                if (view !== ViewType.PERSPECTIVE) {
```
Done

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -104,18 +121,166 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+            const { x, y, z } = this.cube.perspective.position;
+            // @ts-ignore
+            const { width, height, depth } = this.cube.perspective.geometry.parameters;
+            // @ts-ignore
```
Done

jayraj-solanki commented on cvat-canvas3d/src/typescript/canvas3dView.ts:
```diff
@@ -104,18 +121,166 @@ export class Canvas3dViewImpl implements Canvas3dView, Listener {
+                            shapeType: 'cuboid',
+                            frame: this.model.data.imageID,
+                            points,
+                            zOrder: 0 || 0,
```
Removed

jayraj-solanki commented on cvat-ui/src/components/annotation-page/top-bar/statistics-modal.tsx:
```diff
@@ -24,13 +26,25 @@ interface Props {
+        canvasInstance,
     } = props;
 
+    const is2D = canvasInstance instanceof Canvas;
```
Getting by instances is faster, so implemented as such

jayraj-solanki commented on cvat-ui/src/components/annotation-page/appearance-block.tsx:
```diff
@@ -144,8 +151,11 @@ function AppearanceBlock(props: Props): JSX.Element {
+        canvasInstance,
     } = props;
 
+    const is2D = canvasInstance instanceof Canvas;
```
Getting by instances is faster, so implemented as such

bsekachev replied:
Why is it faster? Did you conduct any experiments? It makes the code looking more complex and less obvious and it is why better to get dimension from the storage.

bsekachev replied:
And what is more we do need (and do pass) canvasInstance in some components only for the purpose to make this instanceof comparison. I think this is very controversial solution.

manasars replied:
In this case there are only 2 possibilities of task, 2D and 3D which constitutes canvas and canvas3d.

These are easily accessible data structures across the code as we have defined its type as well, thus getting 'instance of' is a quick access , also in chrome engine the instance of has no effects on performance as well.

bsekachev replied:
@manasars 

The question is not about performance. I believe the both ways are fast enough, but the question is about code readability and clarity.
From this POV it is better to get dimension from storage instead of trying getting it from instanceof operator, which is not very clear. 

In other ways appearance block (and other components in the review) should know NOTHING about the canvas, we do not want this API depend on canvas. This is extra, unnecessary link in the code.

jayraj-solanki commented on cvat-ui/src/containers/annotation-page/top-bar/statistics-modal.tsx:
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2020 Intel Corporation
+// Copyright (C) 2021 Intel Corporation
```
Done

jayraj-solanki commented on cvat-ui/src/containers/annotation-page/top-bar/statistics-modal.tsx:
```diff
@@ -40,6 +43,7 @@ function mapStateToProps(state: CombinedState): StateToProps {
         jobInstance,
         jobStatus,
         savingJobStatus,
+        canvasInstance,
```
Fixed

jayraj-solanki commented on cvat-ui/src/components/annotation-page/top-bar/statistics-modal.tsx:
```diff
@@ -138,6 +152,29 @@ export default function StatisticsModalComponent(props: Props): JSX.Element {
+            title: <Text strong> Total </Text>,
+            dataIndex: 'total',
+            key: 'total',
+        },
```
Done

jayraj-solanki commented on cvat-ui/src/components/annotation-page/top-bar/right-group.tsx:
```diff
@@ -152,6 +153,9 @@ function RightGroup(props: Props): JSX.Element {
                         }
+                        if (canvasInstance instanceof Canvas3d) {
+                            canvasInstance.fit();
+                        }
```
Fixed

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-menu.tsx:
```diff
@@ -227,23 +230,25 @@ function RemoveItem(props: ItemProps): JSX.Element {
+        readonly, shapeType, objectType, colorBy, canvasInstance,
     } = props;
 
+    const is2D = canvasInstance instanceof Canvas;
```
Getting by instances is faster, so implemented as such

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/styles.scss:
```diff
@@ -13,7 +13,7 @@
     background: $border-color-3;
     top: $grid-unit-size;
-    right: $grid-unit-size;
+    right: $grid-unit-size*5;
```
The earlier code had collision with top-side bar button, so increased to 5x

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/controls-side-bar/controls-side-bar.tsx:
```diff
@@ -5,37 +5,112 @@
-            <MoveControl canvasInstance={canvasInstance} activeControl={activeControl} />
-            <CursorControl
+            <GlobalHotKeys keyMap={subKeyMap} handlers={handlers} />
+            <ObservedMoveControl canvasInstance={canvasInstance} activeControl={activeControl} />
```
Fixed

jayraj-solanki commented on cvat-ui/src/components/annotation-page/standard3D-workspace/standard3D-workspace.tsx:
```diff
@@ -8,12 +8,21 @@ import Layout from 'antd/lib/layout';
         <Layout hasSider className='cvat-standard-workspace'>
             <ControlsSideBarContainer />
             <CanvasWrapperContainer />
+            <ObjectSideBarComponent objectsList={<ObjectsListContainer />} />
```
Fixed

manasars commented on cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:
```diff
@@ -10,14 +10,23 @@ import {
 import ContextImage from '../standard3D-workspace/context-image/context-image';
 import CVATTooltip from '../../common/cvat-tooltip';
+import { LogType } from '../../../cvat-logger';
+import getCore from '../../../cvat-core-wrapper';
```
Resolved

manasars commented on cvat-ui/src/components/annotation-page/canvas/canvas-wrapper3D.tsx:
```diff
@@ -10,14 +10,23 @@ import {
-} from 'cvat-canvas3d-wrapper';
+    ColorBy, ContextMenuType, ObjectType, Workspace,
+} from 'reducers/interfaces';
+import { CAMERA_ACTION, Canvas3d, ViewType } from 'cvat-canvas3d-wrapper';
```
Resolved

manasars commented on cvat-canvas3d/src/typescript/cuboid.ts:
```diff
@@ -2,19 +2,139 @@
+    switch (viewType) {
+        case 'top':
+            // @ts-ignore
+            rotationHelper.position.set(instance.geometry.parameters.height / 2 + 0.1,
```
Its the rotation helper placement in cuboid, added naming convention accordingly

bsekachev commented:
An issue with cuboid editing. It gets the initial position after I released my mouse cursor

![Peek 2021-05-13 09-26](https://user-images.githubusercontent.com/40690378/118086859-4c051080-b3cd-11eb-8bb8-dad37795aa5f.gif)

bsekachev commented:
Still opened issues:

- [ ] Redraw shape does not work correctly. Now it started drawing, but instead of redrawing creates one more shape
![Peek 2021-05-13 09-55](https://user-images.githubusercontent.com/40690378/118089757-63de9380-b3d1-11eb-9fed-4e293d0dbaad.gif)
- [ ] Still opened question with fitting by doubleclick. I would vote for keeping default behaviour (fit by dblclick). @nmanovic  your ideas?
- [ ] Cuboids keep being white after they are not hovered anymore
![Peek 2021-05-13 10-00](https://user-images.githubusercontent.com/40690378/118090286-10207a00-b3d2-11eb-910a-7d4b36a7a28a.gif)

> One possible solution could be implementing inhouse compiled version of Open3d in headless mode which is is not a straightforward approach.

I suggested some ideas with using context images and 3D tag, but need to decide collectively the final solution and implement it. 

> The edge color of cuboid will coincide with White, let us know if it's fine.

Yes, looks much better, thanks.

manasars commented:
> An issue with cuboid editing. It gets the initial position after I released my mouse cursor
> 
> ![Peek 2021-05-13 09-26](https://user-images.githubusercontent.com/40690378/118086859-4c051080-b3cd-11eb-8bb8-dad37795aa5f.gif)

Fixed the issue

manasars commented:
> Still opened issues:
> 
> * [ x]  Redraw shape does not work correctly. Now it started drawing, but instead of redrawing creates one more shape
>   ![Peek 2021-05-13 09-55](https://user-images.githubusercontent.com/40690378/118089757-63de9380-b3d1-11eb-9fed-4e293d0dbaad.gif)
In case of 3d redraw shape will work like re-place shape since here we do not draw like 2D, so as per this behaviour it works as expected. On click of Shift+N temporarily hides the existing cuboid (same as 2d) and places another cuboid .
> * [x ]  Still opened question with fitting by doubleclick. I would vote for keeping default behaviour (fit by dblclick). @nmanovic  your ideas?
We have enabled double click to now re-size to origin when not in DRAW mode.
> * [ x]  Cuboids keep being white after they are not hovered anymore
>   ![Peek 2021-05-13 10-00](https://user-images.githubusercontent.com/40690378/118090286-10207a00-b3d2-11eb-910a-7d4b36a7a28a.gif)
> Resolved
> > One possible solution could be implementing inhouse compiled version of Open3d in headless mode which is is not a straightforward approach.
> 
> I suggested some ideas with using context images and 3D tag, but need to decide collectively the final solution and implement it.
> In cases where we do not have context-images(as it is not mandatory), what do we display?

bsekachev commented:
@ActiveChooN  Do you have any comments on the PR?

bsekachev commented:
@manasars 
> On click of Shift+N temporarily hides the existing cuboid (same as 2d) and places another cuboid.

Why do you temporary hide the previous cuboid? It should be removed IMHO, because this is redraw function. Redraw means that the previous cuboid is removed and a new one is created. In your case what is difference between draw and redraw?

- [ ] One more issue found is with copy paste. After I copy modified cuboid (with non-standard form), a cuboid with standard form is being pasted. 

- [ ] When we select move control, activated object should be reset. 

- [ ] One more not standard behavior I noticed is that there is a difference in object selection algorithm.
In 2D you need to hover an object on the canvas to select it (after that you can lock it, hide, it is highlighted in sidebar). In 3D you must click it on canvas.
At the same time in 2D when you hover an object on the sidebar it also activates the object and it work the same way in 3D.
Could you please explain the behavior? I believe it should work the same way regardless off task dimension.

I believe occluded objects should have dotted outlined borders also at the top/front/bottom views.  Do not you think so?

bsekachev commented:
Strange behavior I noticed twice. 
Objects are shifted after activation. Steps to reproduce are not clear. It just appears after some time and different actions on the tab. 

![Peek 2021-05-18 12-21](https://user-images.githubusercontent.com/40690378/118626524-9e7e6c80-b7d3-11eb-8164-13b0dfd16c48.gif)

bsekachev commented:
When we have one object within another object. The smaller object should be activatable on canvas when we hover it. Please, check the way it works for 2D case, should be implemented similar way. 
Now we can't activate the smallest object from canvas, only the external.



![Screenshot from 2021-05-18 12-24-43](https://user-images.githubusercontent.com/40690378/118627053-0f258900-b7d4-11eb-9f27-d4eafb12b6c3.png)

bsekachev commented:
Size of controlling points should be independent of scaling. The issue is demonstrated on the screenshot below:

![Screenshot from 2021-05-18 13-31-01](https://user-images.githubusercontent.com/40690378/118636699-a216f100-b7dd-11eb-8d80-867a6a6f4a12.png)

When I start moving a shape, then up cursor to the top border, release mouse cursor, the shape on the view continue moving after the cursor released:
![Peek 2021-05-18 13-33](https://user-images.githubusercontent.com/40690378/118636901-ddb1bb00-b7dd-11eb-8cf8-5bfd690ee129.gif)

ActiveChooN commented:
@manasars
- Got this error when trying to create a cuboid with [this](https://github.com/PointCloudLibrary/data/tree/master/segmentation/mOSD/learn) dataset
```
three.module.js:9997 THREE.BufferGeometry.computeBoundingSphere(): Computed radius is NaN. The "position" attribute is likely to have NaN values. BufferGeometry
```
- Looks like a bug with cuboid resizing (for the note: I've rotated cuboid a few times)

https://user-images.githubusercontent.com/8168727/118636097-fe2d4580-b7dc-11eb-8592-111fef5fe3c9.mp4

- Ghost cuboids are appearing after multiple `N` key pressing

https://user-images.githubusercontent.com/8168727/118637397-6c263c80-b7de-11eb-9ceb-844e1d1e8282.mp4

manasars commented:
> @manasars
> 
> * Got this error when trying to create a cuboid with [this](https://github.com/PointCloudLibrary/data/tree/master/segmentation/mOSD/learn) dataset
> 
> ```
> three.module.js:9997 THREE.BufferGeometry.computeBoundingSphere(): Computed radius is NaN. The "position" attribute is likely to have NaN values. BufferGeometry
> ```
> 
> * Looks like a bug with cuboid resizing (for the note: I've rotated cuboid a few times)
> 
>  Computer.Vision.Annotation.Tool.-.Google.Chrome.2021-05-18.13-22-22.mp4 
> * Ghost cuboids are appearing after multiple `N` key pressing
> 
>  Computer.Vision.Annotation.Tool.-.Google.Chrome.2021-05-18.13-37-37.mp4

Hi @ActiveChooN 
Could you kindly try with other dataset, for this particular one looks like an improper/corrupt file (mentioned to Boris earlier) and it does not work in Supervisely as well.

ActiveChooN commented:
> Could you kindly try with other dataset, for this particular one looks like an improper/corrupt file (mentioned to Boris earlier) and it does not work in Supervisely as well.

Sure, I will test with another one.

dvkruchinin commented:
> When I start moving a shape, then up cursor to the top border, release mouse cursor, the shape on the view continue moving after the cursor released

This is also reproduced when click outside the shape (top/side/front). Also strange behavior when deactivating the shape. I can also interact with the shape on top/side/front at this time. 
![Animation](https://user-images.githubusercontent.com/33020454/118652140-f4144280-b7ee-11eb-87de-88cc414a8d84.gif)

dvkruchinin commented:
The shape on the sideview temporarily turns into 3d when move the shape above the upper border, click on the free space of the 3d perspective to deactivate the object and then activate it.
This is reproduce when move the shape to the side/front view. But the display shape of the shape changes only on the side view
![canvas3d_2-issue](https://user-images.githubusercontent.com/33020454/118845614-9a864380-b8d4-11eb-94ab-046ad4abea15.gif)

dvkruchinin commented:
If the shape on the top/side/front view is "taken" by the edge with the mouse and moved, the shape moves so that the cursor appears in the center of the shape. This probably shouldn't happen.
![canvas3d_3-issue](https://user-images.githubusercontent.com/33020454/118926676-64cf7200-b949-11eb-9afb-4f1ab9c17d7a.gif)

dvkruchinin commented:
It is probably worth redesigning the drawing of the image taking into account the borders of the selected shape. I also think it would be more convenient when adding a lower border for the top/side/front view.
![canvas3d_4-issue](https://user-images.githubusercontent.com/33020454/118932226-7700de80-b950-11eb-98c4-6f385705082e.gif)

dvkruchinin commented:
Under the scenario:
1. Create a cuboid
2. Click on any view from top/side/front outside the shape
3. Switch lock property or press L
Got an error:
```
three.module.js:44208 Uncaught TypeError: Cannot read property 'layers' of undefined
    at DD (three.module.js:44208)
    at AD.intersectObjects (three.module.js:44285)
    at FI.initiateAction (canvas3dView.ts:1407)
    at canvas3dView.ts:857
    at Array.forEach (<anonymous>)
    at FI.render (canvas3dView.ts:839)
    at VI.render (canvas3d.ts:59)
    at N (canvas-wrapper3D.tsx:213)
```
![canvas3d_5-issue](https://user-images.githubusercontent.com/33020454/118940175-1f1aa580-b959-11eb-8aa7-43f30e07428e.gif)
After that, all the views hang.

dvkruchinin commented:
When I change the label for the cube and copy it, the label that the original one was created with is applied to the copy. Please see the demonstration. 
![canvas3d_6-issue](https://user-images.githubusercontent.com/33020454/118959637-2fd41700-b96b-11eb-80d0-382bda98dd21.gif)
It is reproduce when using the keyboard shortcut Ctrl+C, Ctrl+V or the "Make a copy".

dvkruchinin commented:
Unexpected behavior when drawing a cuboid. Draw a cuboid, press and hold N.
![canvas_3d_7](https://user-images.githubusercontent.com/33020454/118995382-c5819d80-b98f-11eb-8d46-27856591f9bb.gif)

bsekachev commented:
### Summary about found issues, just for convenience:

- [ ] [Code issue with ``instanceof``](https://github.com/openvinotoolkit/cvat/pull/3079#discussion_r634164472)
- [ ] 3D Task preview, opened question how to implement better

>  In cases where we do not have context-images(as it is not mandatory), what do we display?

This is the question, I do not know the right answer. Maybe default image, maybe something else. 

- [ ] [Redrawing a shape does not work correctly when press N multiple times or hold it](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-845167681)
- [ ] [Holding N during drawing also creates ghosts cuboids](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-845167681)
 - [ ] [Redrawing creates one more cuboid](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-840364038)
 - [ ] [Issue with copy/paste. After I copy modified cuboid (with non-standard form), a cuboid with standard form and label is being pasted.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844940603)
 - [ ] When we select move control, an activated object should be reset.
 - [ ] [Difference in selection algorithm](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-842967991)
 - [ ] I believe occluded objects should have dotted outlined borders also at the top/front/bottom views. 
 - [ ] [Issue with shapes shifting after activation](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843008080)
 - [ ] [Size of controlling points should be independent of scaling.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843057301)
 - [ ] [A shape sticks to cursor in some cases](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843057301) and [another case](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843135606)
 - [ ] [Issue with cuboid resizing](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843060092)
 - [ ] [Interaction with the shape is possible using side views after it was deactivated](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843135606)
 - [ ] [Strange behavior when interacting with the "Move the image" button and centering the image.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844182250) @dvkruchinin, I am not sure I can reproduce it. 
 - [ ] [A shape on sideviews gets 3D perspective](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844247049)
 - [ ] [If the shape on the top/side/front view is "taken" by the edge with the mouse and moved, the shape moves so that the cursor appears in the center of the shape.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844724540)
 - [ ] [Uncaught TypeError: Cannot read property 'layers' of undefined](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844825069)
 - [ ] [Issue from @dvkruchinin. Dmitry, could you please give us additional explanation? I am not sure I understand the problem](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844760608)

I hope I mentioned all the considered issues. @ActiveChooN, @dvkruchinin please, check in advance.

dvkruchinin commented:
@bsekachev 
Yes, all my comments are present in this list. Thanks.

> Dmitry, could you please give us additional explanation? I am not sure I understand the problem

Here I meant that it would be more convenient if the drawing of the image was taking into account the boundaries of the selected cuboid. If the cuboid is too large, the borders may not be visible and you will have to scroll.

dvkruchinin commented:
@bsekachev 
>I am not sure I can reproduce it.

A slightly different scenario. But it also depends on the interaction with the "Move the image" button.
1. Create a cuboid.
2. Activate the cuboid.
3. Click on "Move the image" button.
4. Click on cuboid on top or side or front view.
5. Deactivate the cuboid on 3d perspective view. The display of the cuboid on the top/side/front view does not disappear.
![canvas_3d_9](https://user-images.githubusercontent.com/33020454/119085279-1f27ad80-ba0c-11eb-949c-80ba6e966868.gif)

manasars commented:
Hi @bsekachev 
> ### Summary about found issues, just for convenience:
> * [x]  [Code issue with `instanceof`](https://github.com/openvinotoolkit/cvat/pull/3079#discussion_r634164472)
> * []  3D Task preview, opened question how to implement better
> 
> > In cases where we do not have context-images(as it is not mandatory), what do we display?
> 
> This is the question, I do not know the right answer. Maybe default image, maybe something else.
> @Boris at the moment I think we can keep the existing default thumbnail image for preview for all 3D datasets.
> * [x]  [Redrawing a shape does not work correctly when press N multiple times or hold it](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-845167681)
> * [x]  [Holding N during drawing also creates ghosts cuboids](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-845167681)
> * [x]  [Redrawing creates one more cuboid](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-840364038)
Redraw functionality works same as how 2D works.
> * [x]  [Issue with copy/paste. After I copy modified cuboid (with non-standard form), a cuboid with standard form and label is being pasted.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844940603)
> * [x]  When we select move control, an activated object should be reset.
> * [x]  [Difference in selection algorithm](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-842967991)
> * []  I believe occluded objects should have dotted outlined borders also at the top/front/bottom views.
IMHO, I feel having outlined borders in only 3D view is better since the other views are mainly used for translate/resize/rotate and we need not show outlined borders here.
> * [x]  [Issue with shapes shifting after activation](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843008080)
> * [x]  [Size of controlling points should be independent of scaling.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843057301)
> * [x]  [A shape sticks to cursor in some cases](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843057301) and [another case](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843135606)
> * [x]  [Issue with cuboid resizing](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843060092)
> * [x]  [Interaction with the shape is possible using side views after it was deactivated](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-843135606)
> * [x]  [Strange behavior when interacting with the "Move the image" button and centering the image.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844182250) @dvkruchinin, I am not sure I can reproduce it.
> * [x]  [A shape on sideviews gets 3D perspective](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844247049)
> * [x]  [If the shape on the top/side/front view is "taken" by the edge with the mouse and moved, the shape moves so that the cursor appears in the center of the shape.](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844724540)
> * [x]  [Uncaught TypeError: Cannot read property 'layers' of undefined](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844825069)
> * [x]  [Issue from @dvkruchinin](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-844760608). Dmitry, could you please give us additional explanation? I am not sure I understand the problem
> 

Hi @bsekachev , we have fixed all the mentioned issues and commented where-ever necessary, I hope we are good with Milestone-5.

bsekachev commented:
@manasars 

> Hi @bsekachev , we have fixed all the mentioned issues and commented where-ever necessary, I hope we are good with Milestone-5.

Thank you. Let me one more check through the list.

bsekachev commented:
@manasars 

At least I see some conflicts in the PR. Could you resolve them?

manasars commented:
> @manasars
> 
> At least I see some conflicts in the PR. Could you resolve them?

Hi @bsekachev , we are checking the same.
Last when we tried to commit changes, prettier was failing on latest develop changes -> i.e site/content/en/_index.html

bsekachev commented:
@manasars 

>  Last when we tried to commit changes, prettier was failing on latest develop changes -> i.e site/content/en/_index.html

Please, try to add ``site/`` to ``/.prettierignore`` before merge. Let me know if it does not help.

manasars commented:
> @manasars
> 
> At least I see some conflicts in the PR. Could you resolve them?

Hi @bsekachev , we have resolved all the conflicts.
Additionally we have also implemented the Group feature which was discussed earlier with Nikita.

bsekachev commented:
@manasars 

> The mentioned test cases are failing because of the change in selection algorithm(#3079 (comment)) as a part of review changes.
Now, the cuboid is selected on mouse hover and gets activated in Top/Side/Front views on hover same as in 2D.

Another test fails because of the fact that the "Group control" has been added. 

@dvkruchinin Could you please update the tests?

bsekachev commented:
@manasars 

- [ ] Code issue with instanceof (still was not fixed, quick search shows at least several files)
```
/cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx
/cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-menu.tsx
/cvat-ui/src/components/annotation-page/top-bar/statistics-modal.tsx
```

 - [ ] [Redrawing creates one more cuboid](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-840364038) (Still reproducable. It does not work like in 2D. In 2D when you redraw an object, the previous is deleted. In 3D not)
 
While testing again found several new problems:
- [ ] Pressing H while rotation leads to strange behavior
![Peek 2021-05-25 11-07](https://user-images.githubusercontent.com/40690378/119463306-52d74000-bd4a-11eb-8f3c-d8b74f5d7f86.gif)
- [ ] Pressing H while resizing leads to other strange behavior
![Peek 2021-05-25 11-08](https://user-images.githubusercontent.com/40690378/119463388-65517980-bd4a-11eb-82ff-8355e33ee1d4.gif)
- [ ] Pressing Q while resizing/rotation also leads to other strange behavior
- [ ] When we keep hold ``ctrl`` an activated object should not be changed. It is one of basic functions in 2D and 3D totally lacks of it. 
- [ ] Review mode does not work at all (a job with review status cannot be opened). Let me check SOW additionally to find out if this was mentioned. 
- [ ] The shape gets white while moving
![Peek 2021-05-25 11-18](https://user-images.githubusercontent.com/40690378/119463962-f32d6480-bd4a-11eb-8ef9-5a5722068f29.gif)

dvkruchinin commented:
@bsekachev 
> Could you please update the tests?

I have prepared a PR in which I updated the existing tests and add new ones for this functionality: #3228. Unfortunately, I was not able to create a PR for the current PR.

@manasars 
Can you please take the necessary test changes from the specified PR? (case56, case64).

bsekachev commented:
Hi @manasars 

I read in SOW and there is nothing about review mode, so it does not obligatory to support it for 3D. 
But I believe we should remove item "Request a review" in the annotation menu then, because once a job moved to "Review" state, it cannot be more opened anyhow in UI (errors occur) and also we cannot change the state anyhow. The job become "broken".

dvkruchinin commented:
Hi @manasars 
Have you disabled changing the color of the cuboid to 3D perspective view when hover the cursor over it? Previously, the cuboid turned gray. Therefore, the test fails. If so then I need to adapt the test for this.

manasars commented:
> https://github.com/openvinotoolkit/cvat/pull/3228/files#diff-7829468e86c1cc5d5133195b5cb48e1ff6c75e3e9203777f6b2e379d9e4882b3

Hi @dvkruchinin ,

Yes, we have removed the color change to grey on hover to be compatible with 2D and as per one of the review comments.
Ok, I shall integrate the changes from main.yml file as well into the PR.

manasars commented:
> Hi @manasars
> 
> I read in SOW and there is nothing about review mode, so it does not obligatory to support it for 3D.
> But I believe we should remove item "Request a review" in the annotation menu then, because once a job moved to "Review" state, it cannot be more opened anyhow in UI (errors occur) and also we cannot change the state anyhow. The job become "broken".

Yes Boris, we shall remove/hide the item in annotation menu.

jayraj-solanki commented:
hi @bsekachev 
> 
> * [x]  Code issue with instanceof (still was not fixed, quick search shows at least several files)
> 
> ```
> /cvat-ui/src/components/annotation-page/standard-workspace/controls-side-bar/draw-shape-popover.tsx
> /cvat-ui/src/components/annotation-page/standard-workspace/objects-side-bar/object-item-menu.tsx
> /cvat-ui/src/components/annotation-page/top-bar/statistics-modal.tsx
> ```
>
>Fixed

> * [x]  [Redrawing creates one more cuboid](https://github.com/openvinotoolkit/cvat/pull/3079#issuecomment-840364038) (Still reproducable. It does not work like in 2D. In 2D when you redraw an object, the previous is deleted. In 3D not)
>
>In one of test MAC system it was working as creating 2 cuboids in 2d as well and in Linux system is deleting the older cube and placing the new one. We fixed it now

> * [x]  Pressing H while rotation leads to strange behavior
> * [x]  Pressing H while resizing leads to other strange behavior
>
>We have fixed the above two issues, but in 3D once the mouse is released after any edit operation(resize, move or rotate) then the hide function can be seen

> * [x]  Pressing Q while resizing/rotation also leads to other strange behavior
>
>We have fixed the above issue, but in 3D once the mouse is released after any edit operation(resize, move or rotate) then the occlusion function can be seen

> * [x]  When we keep hold `ctrl` an activated object should not be changed. It is one of basic functions in 2D and 3D totally lacks of it.
>
>Fixed

> * [x]  The shape gets white while moving
>
>Fixed

> * [ ]  Review mode does not work at all (a job with review status cannot be opened). Let me check SOW additionally to find out if this was mentioned.
>
>Out of Scope

>* [x] I read in SOW and there is nothing about review mode, so it does not obligatory to support it for 3D.
>But I believe we should remove item "Request a review" in the annotation menu then, because once a 
>job moved to "Review" state, it cannot be more opened anyhow in UI (errors occur) and also we cannot
> change the state anyhow. The job become "broken".
>
>Pushed changes

bsekachev commented:
One of tests is still failed. If @dvkruchinin has provided fixed test in his PR #3228 could you fetch it from there?

manasars commented:
> One of tests is still failed. If @dvkruchinin has provided fixed test in his PR #3228 could you fetch it from there?

@bsekachev , I checked there are a lot of files changes and even method changes in the 3228 branch now, not sure if I will be able to merge correctly.

dvkruchinin commented:
@manasars 
> Kindly correct me if my understanding is wrong.

You did everything right, but in my PR, I also do test refactoring. Please pull up the changes from file https://github.com/openvinotoolkit/cvat/pull/3228/files#diff-6f73123ddf0e3c11c9d6d279110a1d5bca2b1f57e0d527afb0c7c909dab8f19b

manasars commented:
Thank you @dvkruchinin ,

@bsekachev : Updated the test which was failing, all tests have passed.
